I'm missing a utility from another PR;

```
integration/image/prune_test.go:147:45: undefined: image.InspectResponse
integration/image/prune_test.go:151:47: undefined: image.InspectResponse
integration/image/prune_test.go:157:47: undefined: image.InspectResponse
integration/image/prune_test.go:163:47: undefined: image.InspectResponse
integration/image/prune_test.go:169:47: undefined: image.InspectResponse
integration/image/prune_test.go:177:47: undefined: image.InspectResponse
integration/image/prune_test.go:185:47: undefined: image.InspectResponse
```

--- a/daemon/containerd/image_prune.go
+++ b/daemon/containerd/image_prune.go
@@ -2,6 +2,7 @@ package containerd
 
 import (
 	"context"
+	"sort"
 	"strings"
 
 	containerdimages "github.com/containerd/containerd/images"
@@ -107,8 +108,16 @@ func (i *ImageService) pruneUnused(ctx context.Context, filterFunc imageFilterFu
 
 	usedDigests := filterImagesUsedByContainers(ctx, i.containers.List(), imagesToPrune)
 
+	// Sort images by name to make the behavior deterministic and consistent with graphdrivers.
+	sorted := make([]string, 0, len(imagesToPrune))
+	for name := range imagesToPrune {
+		sorted = append(sorted, name)
+	}
+	sort.Strings(sorted)
+
 	// Make sure we don't delete the last image of a particular digest used by any container.
-	for name, img := range imagesToPrune {
+	for _, name := range sorted {
+		img := imagesToPrune[name]
 		dgst := img.Target.Digest
 
 		if digestRefCount[dgst] > 1 {

--- a/daemon/containerd/image_prune.go
+++ b/daemon/containerd/image_prune.go
@@ -5,6 +5,7 @@ import (
 	"strings"
 
 	containerdimages "github.com/containerd/containerd/images"
+	"github.com/containerd/containerd/tracing"
 	cerrdefs "github.com/containerd/errdefs"
 	"github.com/containerd/log"
 	"github.com/distribution/reference"
@@ -75,6 +76,10 @@ func (i *ImageService) ImagesPrune(ctx context.Context, fltrs filters.Args) (*im
 // In case a digested and tagged reference was used (e.g. `docker run alpine:latest@sha256:82d1e9d7ed48a7523bdebc18cf6290bdb97b82302a8a9c27d4fe885949ea94d1`),
 // the alpine:latest image will be kept.
 func (i *ImageService) pruneUnused(ctx context.Context, filterFunc imageFilterFunc, danglingOnly bool) (*image.PruneReport, error) {
+	ctx, span := tracing.StartSpan(ctx, "ImageService.pruneUnused")
+	span.SetAttributes(tracing.Attribute("danglingOnly", danglingOnly))
+	defer span.End()
+
 	allImages, err := i.images.List(ctx)
 	if err != nil {
 		return nil, err
@@ -125,6 +130,10 @@ func filterImagesUsedByContainers(ctx context.Context,
 	allContainers []*container.Container,
 	imagesToPrune map[string]containerdimages.Image,
 ) (usedDigests map[digest.Digest]struct{}) {
+	ctx, span := tracing.StartSpan(ctx, "filterImagesUsedByContainers")
+	span.SetAttributes(tracing.Attribute("count", len(allContainers)))
+	defer span.End()
+
 	// Image specified by digests that are used by containers.
 	usedDigests = map[digest.Digest]struct{}{}
 
@@ -178,6 +187,10 @@ func filterImagesUsedByContainers(ctx context.Context,
 // pruneAll deletes all images in the imagesToPrune map.
 func (i *ImageService) pruneAll(ctx context.Context, imagesToPrune map[string]containerdimages.Image) (*image.PruneReport, error) {
 	report := image.PruneReport{}
+
+	ctx, span := tracing.StartSpan(ctx, "ImageService.pruneAll")
+	span.SetAttributes(tracing.Attribute("count", len(imagesToPrune)))
+	defer span.End()
 
 	possiblyDeletedConfigs := map[digest.Digest]struct{}{}
 	var errs error

--- a/integration/image/prune_test.go
+++ b/integration/image/prune_test.go
@@ -4,6 +4,7 @@ import (
 	"strings"
 	"testing"
 
+	"github.com/docker/docker/api/types"
 	containertypes "github.com/docker/docker/api/types/container"
 	"github.com/docker/docker/api/types/filters"
 	"github.com/docker/docker/api/types/image"
@@ -98,29 +99,29 @@ func TestPruneDontDeleteUsedImage(t *testing.T) {
 	} {
 		for _, tc := range []struct {
 			name    string
-			imageID func(t *testing.T, inspect image.InspectResponse) string
+			imageID func(t *testing.T, inspect types.ImageInspect) string
 		}{
 			{
 				name: "full id",
-				imageID: func(t *testing.T, inspect image.InspectResponse) string {
+				imageID: func(t *testing.T, inspect types.ImageInspect) string {
 					return inspect.ID
 				},
 			},
 			{
 				name: "full id without sha256 prefix",
-				imageID: func(t *testing.T, inspect image.InspectResponse) string {
+				imageID: func(t *testing.T, inspect types.ImageInspect) string {
 					return strings.TrimPrefix(inspect.ID, "sha256:")
 				},
 			},
 			{
 				name: "truncated id (without sha256 prefix)",
-				imageID: func(t *testing.T, inspect image.InspectResponse) string {
+				imageID: func(t *testing.T, inspect types.ImageInspect) string {
 					return strings.TrimPrefix(inspect.ID, "sha256:")[:8]
 				},
 			},
 			{
 				name: "repo and digest without tag",
-				imageID: func(t *testing.T, inspect image.InspectResponse) string {
+				imageID: func(t *testing.T, inspect types.ImageInspect) string {
 					skip.If(t, !testEnv.UsingSnapshotter())
 
 					return "busybox@" + inspect.ID
@@ -128,7 +129,7 @@ func TestPruneDontDeleteUsedImage(t *testing.T) {
 			},
 			{
 				name: "tagged and digested",
-				imageID: func(t *testing.T, inspect image.InspectResponse) string {
+				imageID: func(t *testing.T, inspect types.ImageInspect) string {
 					skip.If(t, !testEnv.UsingSnapshotter())
 
 					return "busybox:latest@" + inspect.ID
@@ -136,7 +137,7 @@ func TestPruneDontDeleteUsedImage(t *testing.T) {
 			},
 			{
 				name: "repo digest",
-				imageID: func(t *testing.T, inspect image.InspectResponse) string {
+				imageID: func(t *testing.T, inspect types.ImageInspect) string {
 					// graphdriver won't have a repo digest
 					skip.If(t, len(inspect.RepoDigests) == 0, "no repo digest")
 

--- a/integration/image/prune_test.go
+++ b/integration/image/prune_test.go
@@ -54,6 +54,51 @@ func TestPruneDontDeleteUsedDangling(t *testing.T) {
 	assert.NilError(t, err, "Test dangling image should still exist")
 }
 
+func TestPruneLexographicalOrder(t *testing.T) {
+	skip.If(t, testEnv.DaemonInfo.OSType == "windows", "cannot start multiple daemons on windows")
+	skip.If(t, testEnv.IsRemoteDaemon, "cannot run daemon when remote daemon")
+
+	ctx := setupTest(t)
+
+	d := daemon.New(t)
+	d.Start(t)
+	defer d.Stop(t)
+
+	apiClient := d.NewClientT(t)
+	defer apiClient.Close()
+
+	d.LoadBusybox(ctx, t)
+
+	inspect, _, err := apiClient.ImageInspectWithRaw(ctx, "busybox:latest")
+	assert.NilError(t, err)
+
+	id := inspect.ID
+
+	var tags = []string{"h", "a", "j", "o", "s", "q", "w", "e", "r", "t"}
+	for _, tag := range tags {
+		err = apiClient.ImageTag(ctx, id, "busybox:"+tag)
+		assert.NilError(t, err)
+	}
+	err = apiClient.ImageTag(ctx, id, "busybox:z")
+	assert.NilError(t, err)
+
+	_, err = apiClient.ImageRemove(ctx, "busybox:latest", image.RemoveOptions{Force: true})
+	assert.NilError(t, err)
+
+	// run container
+	cid := container.Create(ctx, t, apiClient, container.WithImage(id))
+	defer container.Remove(ctx, t, apiClient, cid, containertypes.RemoveOptions{Force: true})
+
+	pruned, err := apiClient.ImagesPrune(ctx, filters.NewArgs(filters.Arg("dangling", "false")))
+	assert.NilError(t, err)
+
+	assert.Check(t, is.Len(pruned.ImagesDeleted, len(tags)))
+	for _, p := range pruned.ImagesDeleted {
+		assert.Check(t, is.Equal(p.Deleted, ""))
+		assert.Check(t, p.Untagged != "busybox:z")
+	}
+}
+
 // Regression test for https://github.com/moby/moby/issues/48063
 func TestPruneDontDeleteUsedImage(t *testing.T) {
 	skip.If(t, testEnv.DaemonInfo.OSType == "windows", "cannot start multiple daemons on windows")
@@ -82,18 +127,19 @@ func TestPruneDontDeleteUsedImage(t *testing.T) {
 			// busybox:other tag pointing to the same image.
 			name: "two tags",
 			prepare: func(t *testing.T, d *daemon.Daemon, apiClient *client.Client) error {
-				return apiClient.ImageTag(ctx, "busybox:latest", "busybox:other")
+				return apiClient.ImageTag(ctx, "busybox:latest", "busybox:a")
 			},
 			check: func(t *testing.T, apiClient *client.Client, pruned image.PruneReport) {
 				if assert.Check(t, is.Len(pruned.ImagesDeleted, 1)) {
 					assert.Check(t, is.Equal(pruned.ImagesDeleted[0].Deleted, ""))
-					t.Log("Untagged", pruned.ImagesDeleted[0].Untagged)
+					assert.Check(t, is.Equal(pruned.ImagesDeleted[0].Untagged, "busybox:a"))
 				}
 
-				_, _, err1 := apiClient.ImageInspectWithRaw(ctx, "busybox:latest")
-				_, _, err2 := apiClient.ImageInspectWithRaw(ctx, "busybox:other")
+				_, _, err := apiClient.ImageInspectWithRaw(ctx, "busybox:a")
+				assert.Check(t, err != nil, "Busybox:a image should be deleted")
 
-				assert.Check(t, err1 != err2 && (err1 == nil || err2 == nil), "One of the images should still exist")
+				_, _, err = apiClient.ImageInspectWithRaw(ctx, "busybox:latest")
+				assert.Check(t, err == nil, "Busybox:latest image should still exist")
 			},
 		},
 	} {

